internal/services/ingest.go

```diff
@@ -0,0 +1,231 @@
+
+	if !prepared {
+		err = m.LedgerBackend.PrepareRange(ctx, ledgerRange)
+		if err != nil {
+			return fmt.Errorf("preparing range: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func trackServiceHealth(heartbeat chan any) {
+	const alertAfter = time.Second * 60
+	ticker := time.NewTicker(alertAfter)
+
+	for {
+		select {
+		case <-ticker.C:
+			warn := fmt.Sprintf("ingestion service stale for over %s", alertAfter)
+			log.Warn(warn)
+			// TODO: track in Sentry
+			// sentry.CaptureMessage(warn)
+			ticker.Reset(alertAfter)
+		case <-heartbeat:
+			ticker.Reset(alertAfter)
+		}
+	}
+}
+
+func (m *IngestManager) processLedger(ctx context.Context, ledger uint32, ledgerMeta xdr.LedgerCloseMeta) (err error) {
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(m.NetworkPassphrase, ledgerMeta)
+	if err != nil {
+		return fmt.Errorf("creating ledger reader: %w", err)
+	}
+
+	ledgerCloseTime := time.Unix(int64(ledgerMeta.LedgerHeaderHistoryEntry().Header.ScpValue.CloseTime), 0).UTC()
+	ledgerSequence := ledgerMeta.LedgerSequence()
+
+	dbTx, err := m.PaymentModel.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			rollbackErr := dbTx.Rollback()
+			if rollbackErr != nil {
+				log.Ctx(ctx).Error(rollbackErr)
+			}
+		}
+	}()
+
+	for {
+		tx, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("reading transaction: %w", err)
+		}
+
+		if !tx.Result.Successful() {
+			continue
+		}
+
+		txHash := utils.TransactionHash(ledgerMeta, int(tx.Index))
+		txMemo := utils.Memo(tx.Envelope.Memo(), txHash)
+
+		for idx, op := range tx.Envelope.Operations() {
+			opIdx := idx + 1
+
+			payment := data.Payment{
+				OperationID:     utils.OperationID(int32(ledgerSequence), int32(tx.Index), int32(opIdx)),
+				OperationType:   op.Body.Type.String(),
+				TransactionID:   utils.TransactionID(int32(ledgerSequence), int32(tx.Index)),
+				TransactionHash: txHash,
+				From:            utils.SourceAccount(op, tx),
+				CreatedAt:       ledgerCloseTime,
+				Memo:            utils.SanitizeUTF8(txMemo), // Field is subject to user input
+			}
+
+			switch op.Body.Type {
+			case xdr.OperationTypePayment:
+				fillPayment(&payment, op.Body)
+			case xdr.OperationTypePathPaymentStrictSend:
+				fillPathSend(&payment, op.Body, tx, opIdx)
+			case xdr.OperationTypePathPaymentStrictReceive:
+				fillPathReceive(&payment, op.Body, tx, opIdx)
+			default:
+				continue
+			}
+
+			err = m.PaymentModel.AddPayment(ctx, dbTx, payment)
+			if err != nil {
+				return fmt.Errorf("adding payment for ledger %d, tx %q (%d), operation %d (%d): %w", ledgerSequence, txHash, tx.Index, payment.OperationID, opIdx, err)
+			}
+		}
+	}
+
+	err = m.PaymentModel.UpdateLatestLedgerSynced(ctx, m.LedgerCursorName, ledger)
+	if err != nil {
+		return err
+	}
+
+	err = dbTx.Commit()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func fillPayment(payment *data.Payment, operation xdr.OperationBody) {
```

Comment:
can we test these functions to check if all values are being filled with the corrected values?

Reply:
Added a test to the `processLedger` function, which calls this method internally: [ac1f46e](https://github.com/stellar/wallet-backend/pull/14/commits/ac1f46ea963cc6657b6329ff0fbc4f4d32ee8d26)